gapii/cc/vulkan_mid_execution.cpp

```diff
@@ -153,1690 +156,1780 @@ class TemporaryRenderPass {
 
   ~TemporaryRenderPass() {
     for (auto m : temporary_render_passes_) {
-      spy_->RecreateDestroyRenderPass(observer_, m->mDevice,
-                                        m->mVulkanHandle);
+      spy_->RecreateDestroyRenderPass(observer_, m->mDevice, m->mVulkanHandle);
     }
   }
 
   bool has(VkRenderPass renderpass) {
-      return std::find_if(temporary_render_passes_.begin(), 
-        temporary_render_passes_.end(), 
-        [renderpass](std::shared_ptr<RenderPassObject>& p) {
-             return p->mVulkanHandle == renderpass;
-        }) != temporary_render_passes_.end();
+    return std::find_if(temporary_render_passes_.begin(),
+                        temporary_render_passes_.end(),
+                        [renderpass](std::shared_ptr<RenderPassObject>& p) {
+                          return p->mVulkanHandle == renderpass;
+                        }) != temporary_render_passes_.end();
   }
 
  private:
   CallObserver* observer_;
   VulkanSpy* spy_;
   std::vector<std::shared_ptr<RenderPassObject>> temporary_render_passes_;
 };
+
+template <typename ObjectClass>
+VkDevice getObjectCreatingDevice(std::shared_ptr<ObjectClass> obj) {
+  return obj->mDevice;
+}
+
+template <>
+VkDevice getObjectCreatingDevice<InstanceObject>(
+    std::shared_ptr<InstanceObject> obj) {
+  return VkDevice(0);
+}
+
+
+template <>
+VkDevice getObjectCreatingDevice<PhysicalDeviceObject>(
+    std::shared_ptr<PhysicalDeviceObject> obj) {
+  return VkDevice(0);
+}
+
+template <>
+VkDevice getObjectCreatingDevice<DeviceObject>(
+    std::shared_ptr<DeviceObject> obj) {
+  return obj->mVulkanHandle;
+}
+
+template <>
+VkDevice getObjectCreatingDevice<SurfaceObject>(
+    std::shared_ptr<SurfaceObject> obj) {
+  return VkDevice(0);
+}
+
+template <typename ObjectClass>
+void recreateDebugInfo(VulkanSpy* spy, CallObserver* observer,
+                       uint32_t objectType, std::shared_ptr<ObjectClass> obj) {
```

Comment:
You should probably make this a const std::shared_ptr<>& obj, unless you need to modify the object.
shared_ptr copies are more expensive than we want to pay if we don't have to.

Reply:
Changed to const ref in a83cc08